### PR TITLE
login/v2 implementation for Go

### PR DESCRIPTION
**Pull request by burgerdev**
First draft of the v2 implementation with a fresh API.

### CODE REVIEW

File: go/login/v2/responder.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+type ServerChallenge struct {
+	Message  *Message
+	Response string
+}
+
+// Accept an encoded challenge and produce a response.
+func (r *Responder) Accept(encodedChallenge string) (*ServerChallenge, error) {
+	s := strings.TrimPrefix(encodedChallenge, "/")
+	if len(s) < len(versionPrefix) {
+		return nil, fmt.Errorf("format error")
```
Comment (pkern) [inline go/login/v2/responder.go:58]: I think here it'd be good to at least have two distinct format errors.

File: go/login/v2/responder.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+	var key *glome.PrivateKey
+	ok := false
+	if h.Prefix != nil {
+		key, ok = r.keysByPrefix[*h.Prefix]
+	} else {
+		key, ok = r.keysByIndex[h.Index]
+	}
+	if !ok {
+		return nil, fmt.Errorf("key not found")
```
Comment (pkern) [inline go/login/v2/responder.go:89]: Could we say which key was requested?

File: go/login/v2/README.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Login V2
+
+TODO: describe this package in more detail.
```
[the remark this answers is not in the document]
Comment (burgerdev) [reply go/login/v2/README.md]: Fixed

File: go/login/v2/codec.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+func decodeHandshake(s string) (*handshake, error) {
+	data, err := base64.URLEncoding.DecodeString(s)
+	if err != nil {
+		return nil, err
+	}
+	if len(data) < 33 {
+		return nil, errors.New("handshake too short")
+	}
+
+	handshake := &handshake{}
```
Comment (pkern) [inline go/login/v2/codec.go:102]: I'd weakly prefer if this wouldn't shadow the type.

File: go/login/v2/login_test.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+			}
+			if !reflect.DeepEqual(tc.msg, sc.Message) {
+				t.Errorf("Responder parsed wrong message: got %#v, want %#v", sc.Message, tc.msg)
+			}
+
+			if sc.Response != tc.response {
+				t.Errorf("Responder generated wrong response: got %q, want %q", sc.Response, tc.response)
+			}
+
+			min := uint8(10)
```
Comment (pkern) [inline go/login/v2/login_test.go:228]: You don't need it to be uint8 for slicing, right? So should this just be an int and a cast when assigning it from `*tc.minResponseLength`?

File: go/login/v2/login_test.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+			if tc.minResponseLength != nil {
+				min = *tc.minResponseLength
+			}
+
+			for i := 1; i < int(min); i++ {
```
Comment (pkern) [inline go/login/v2/login_test.go:233]: <=?

**Review by pkern — APPROVED** [review]
Generally LGTM.

File: go/login/v2/login_test.go (hunk 6 above)
Comment (burgerdev) [reply go/login/v2/login_test.go]: No, that part tests that tags smaller than min are rejected. Added a comment for clarity.

**Pull request merged**